### PR TITLE
[v15] refactor(app-shell): use react-apollo v3

### PR DESCRIPTION
**Pull request by emmenko**
Use react-apollo v3, migrate internal queries to `useQuery`, fix tests.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -134,6 +134,10 @@
     "typescript": "3.5.2"
   },
   "resolutions": {
+    "@apollo/react-common": "0.1.0-beta.8",
+    "@apollo/react-components": "0.1.0-beta.7",
+    "@apollo/react-hoc": "0.1.0-beta.7",
+    "@apollo/react-hooks": "0.1.0-beta.10",
```
Comment (emmenko) [inline package.json]: Those resolutions are currently necessary as the beta packages do not reference each other in a proper way (https://github.com/apollographql/react-apollo/blob/8b6abe068c3027f2b539f4906f5ba5d7c628090a/packages/all/package.json#L42-L46)

File: packages/application-shell/src/components/navbar/navbar.js [hunk 2]
```diff
@@ -607,149 +606,100 @@ NavBarLayout.propTypes = {
...
-
-export default flowRight(
-  withRouter, // Connect again, to access the `location` object
-  injectFeatureToggle('projectExtensions', 'areProjectExtensionsEnabled'),
-  injectMenuToggleState,
-  withApplicationsMenu({
-    queryName: 'applicationsMenuQuery',
-    skipRemoteQuery: ownProps => !ownProps.environment.servedByProxy,
-    options: ownProps => ({
+export const NavBar = props => {
```
Comment (emmenko) [inline packages/application-shell/src/components/navbar/navbar.js]: We're using several hooks here and the code is much more readable now!

File: packages/application-shell/src/components/navbar/navbar.js [hunk 3]
```diff
@@ -607,149 +606,100 @@ NavBarLayout.propTypes = {
...
+    useFullRedirectsForLinks: PropTypes.bool.isRequired,
+  }).isRequired,
+  menuVisibilities: PropTypes.objectOf(PropTypes.bool).isRequired,
+  DEV_ONLY__loadNavbarMenuConfig: PropTypes.func,
+  // withRouter
+  location: PropTypes.object.isRequired,
+};
+
+// Use `withRouter` to "connect" again, to access the `location` object
+export default withRouter(NavBar);
```
Comment (emmenko) [inline packages/application-shell/src/components/navbar/navbar.js]: Just one HOC left 🙌

File: packages/application-shell/src/components/user-settings-menu/user-settings-menu.js [hunk 4]
```diff
@@ -87,11 +86,23 @@ const renderLabel = (menu, applicationLanguage) => {
...
-    (props.applicationsMenuQuery &&
-      props.applicationsMenuQuery.applicationsMenu &&
-      props.applicationsMenuQuery.applicationsMenu.appBar) ||
-    [];
+  const applicationsMenu = useApplicationsMenu({
+    queryOptions: {
+      // We can assume here that the navbar already fetched the data, since this
+      // component gets rendered only when the user opens the menu
+      fetchPolicy: 'cache-only',
+      onError: reportErrorToSentry,
```
Comment (emmenko) [inline packages/application-shell/src/components/user-settings-menu/user-settings-menu.js]: This is also interesting and I think we should do that more often. When we don't specifically handle the apollo `error`, we can just bind the error handler with our sentry reporter.

File: packages/application-shell/src/components/user-settings-menu/user-settings-menu.spec.js [hunk 5]
```diff
@@ -83,19 +66,31 @@ describe('rendering', () => {
...
-        );
+      it('should render <UserSettingsMenuBody>', () => {
+        expect(menuStateContainerRenderWrapper).toRender(UserSettingsMenuBody);
       });
     });
   });
 
   describe('<UserSettingsMenuBody>', () => {
     beforeEach(() => {
+      useApplicationsMenu.mockReturnValue({
```
Comment (emmenko) [inline packages/application-shell/src/components/user-settings-menu/user-settings-menu.spec.js]: Easy peasy 🤓

File: packages/application-shell/src/hooks/use-applications-menu/use-applications-menu.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+const defaultApiUrl = window.location.origin;
+
+const defaultConfig = {
+  queryOptions: {},
+  skipRemoteQuery: false,
+  options: {},
+};
+
+const useApplicationsMenu = (config = {}) => {
```
Comment (emmenko) [inline packages/application-shell/src/hooks/use-applications-menu/use-applications-menu.js]: This also got nicer and easier to follow (I hope)

**Comment by emmenko** [comment]
So apparently we need to update to react >=16.9 and testing-library/react@>=8 in order to fix this problem

https://github.com/apollographql/react-apollo/issues/3200#issuecomment-507158053

**Comment by montezume** [comment]
Should we coordinate on making the required test fixes to MC-FE to upgrade to latest RTL there then?

**Comment by emmenko** [comment]
Yeah we need to discuss how we want to proceed here. This PR is not meant to be merged just yet, I just want to experiment and see what we need to do in order to support newer versions.

**Comment by montezume** [comment]
Either way it would be a good idea to update MC FE to latest testing library version... 😄

**Comment by emmenko** [comment]
Definitely! I added a comment in the v14 issue, to keep track of it: https://github.com/commercetools/merchant-center-application-kit/issues/748#issuecomment-507170817

Btw, what was the problem with updating to latest RTL?

**Comment by montezume** [comment]
the breaking changes are listed here

https://github.com/testing-library/dom-testing-library/releases/tag/v4.0.0

getBy:* All getBy and findBy query variants now will throw an error if more than one element is returned. If this is expected, then use getAllBy (or findAllBy) instead.

**Comment by emmenko** [comment]
Now that the new apollo hooks are out, I'm going to pick this up again

**Comment by montezume** [comment]
Maybe we should update to rc version of React 16.9? I had a lot of trouble testing async hooks with the current version...

**Comment by emmenko** [comment]
Yes that as well

**Comment by emmenko** [comment]
Ok we're a bit stuck on this again:

- by not using react@16.9 we get the error for `act` on testing
- by using react@16.9 we get a bunch of errors from react about libraries using the deprecated `componentWillReceiveProps`. So far (according to [this build](https://circleci.com/gh/commercetools/merchant-center-application-kit/5954?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)) the following components need to be fixed:
  - `ConfigureAdapter` (see also https://github.com/tdeekens/flopflip/issues/733)
  - `AutosizeInput` (from `react-input-autosize` used by `react-select`)
  - `Select` (from `react-select`)

File: packages/application-shell/src/hooks/use-applications-menu/use-applications-menu.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        devConfig.menuLoader().then(setMenu);
+      }
+    }
+  }, [getDevConfig, mergedConfig.skipRemoteQuery]);
+
+  // Prepare to fetch config.
+  // We set up the apollo query as lazy, in order to execute it conditionally
+  // since hooks cannot be defined conditionally.
+  const environment = useApplicationContext(context => context.environment);
+  const [loadMenuData, { called, data }] = useLazyQuery(FetchApplicationsMenu, {
```
Comment (emmenko) [inline packages/application-shell/src/hooks/use-applications-menu/use-applications-menu.js]: This is nice!

File: packages/application-shell-connectors/src/components/project-extension-image-regex/project-extension-image-regex.spec.tsx [hunk 8]
```diff
@@ -1,76 +1,148 @@
 import React from 'react';
-import { shallow, ShallowWrapper } from 'enzyme';
-import { ProjectExtensionProviderForImageRegex } from './project-extension-image-regex';
+import {
+  render,
+  RenderResult,
+  waitForElementToBeRemoved,
+} from '@testing-library/react';
```
Comment (emmenko) [inline packages/application-shell-connectors/src/components/project-extension-image-regex/project-extension-image-regex.spec.tsx]: TL;DR: I rewrote the tests to RTL

File: packages/application-shell-connectors/src/components/project-extension-image-regex/project-extension-image-regex.tsx [hunk 9]
```diff
@@ -9,50 +10,52 @@ type TImageRegexOptions = {
...
-        }}
-      >
-        {props.children}
-      </Context.Provider>
-    )}
-  </Query>
-);
+    { target: typeof GRAPHQL_TARGETS.SETTINGS_SERVICE }
+  >(FetchProjectExtensionImageRegex, {
+    onError: reportErrorToSentry,
```
Comment (emmenko) [inline packages/application-shell-connectors/src/components/project-extension-image-regex/project-extension-image-regex.tsx]: Since we're not exposing any graphql error, we can use the `onError` handler to report it to sentry

File: packages/application-shell/src/components/application-shell-provider/application-shell-provider.spec.js [hunk 10]
```diff
@@ -9,7 +9,6 @@ import { setIsAuthenticated } from '../authenticated';
 import { getBrowserLocale } from './utils';
 import ApplicationShellProvider from './application-shell-provider';
 
-jest.mock('@commercetools-frontend/sentry');
```
Comment (emmenko) [inline packages/application-shell/src/components/application-shell-provider/application-shell-provider.spec.js]: It was unnecessary

File: packages/application-shell/src/components/fetch-project/fetch-project.js [hunk 11]
```diff
@@ -145,53 +145,48 @@ export const mapAllDataFencesToGroupedObjectShape = allAppliedDataFences => {
...
-  static propTypes = {
-    projectKey: PropTypes.string,
-    skip: PropTypes.bool,
-    children: PropTypes.func.isRequired,
-  };
-  static defaultProps = {
-    skip: false,
-  };
-  shouldComponentUpdate(nextProps) {
-    return !deepEqual(this.props, nextProps);
```
Comment (emmenko) [inline packages/application-shell/src/components/fetch-project/fetch-project.js]: I couldn't find an useful case for this check, so I left it out when migrating the component to FC.

File: packages/application-shell/src/components/fetch-project/fetch-project.js [hunk 12]
```diff
@@ -145,53 +145,48 @@ export const mapAllDataFencesToGroupedObjectShape = allAppliedDataFences => {
...
-              },
-          })
-        }
-      </Query>
-    );
-  }
-}
+const FetchProject = props => {
+  const { loading, data, error } = useQuery(ProjectQuery, {
+    onError: reportErrorToSentry,
```
Comment (emmenko) [inline packages/application-shell/src/components/fetch-project/fetch-project.js]: In AppShell we don't explicitly handle any `error`, so we report it to sentry in case we get some.

File: packages/application-shell/src/components/fetch-user/fetch-user.js [hunk 13]
```diff
@@ -1,34 +1,24 @@
...
-            error,
-          })
-        }
-      </Query>
-    );
-  }
-}
+const FetchUser = props => {
+  const { loading, data, error } = useQuery(LoggedInUserQuery, {
+    onError: reportErrorToSentry,
```
Comment (emmenko) [inline packages/application-shell/src/components/fetch-user/fetch-user.js]: Same here

File: packages/application-shell/src/components/inject-reducers/inject-reducers.js [hunk 14]
```diff
@@ -24,7 +24,7 @@ export class InjectReducers extends React.PureComponent {
     this.setState({ areReducersInjected: true });
   }
 
-  UNSAFE_componentWillUnmount() {
+  componentWillUnmount() {
```
Comment (emmenko) [inline packages/application-shell/src/components/inject-reducers/inject-reducers.js]: `componentWillUnmount` is not a deprecated lifecycle method. I guess this was confused with `componentWillMount`

File: packages/application-shell/src/components/quick-access/quick-access.js [hunk 15]
```diff
@@ -341,9 +344,11 @@ class QuickAccess extends React.Component {
       <Butler
         intl={this.props.intl}
         historyEntries={this.state.historyEntries}
-        onHistoryEntriesChange={historyEntries =>
-          this.setState({ historyEntries })
-        }
+        onHistoryEntriesChange={historyEntries => {
+          if (!this.isUnmounting) {
```
Comment (emmenko) [inline packages/application-shell/src/components/quick-access/quick-access.js]: Without this check, React throws a 

> Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in

File: packages/l10n/src/create-l10n-injector/create-l10n-injector.tsx [hunk 16]
```diff
@@ -82,17 +82,6 @@ export function createL10NInjector<LoadedData extends {}>({
   propLoadingKey,
   loadLocale,
 }: InjectorOptions<LoadedData>) {
-  if (process.env.NODE_ENV === 'test') {
-    if (React.version.startsWith('16.8')) {
```
Comment (emmenko) [inline packages/l10n/src/create-l10n-injector/create-l10n-injector.tsx]: We can get rid of this now 🎉

**Comment by emmenko** [comment]
Alright, this is now ready for review and merge. I fixed all the issues within the v15 branch, so let's merge this before doing any more work on the v15 branch. /cc @tdeekens

File: packages/application-shell/src/components/fetch-project/fetch-project.js (hunk 12 above)
Comment (tdeekens) [reply packages/application-shell/src/components/fetch-project/fetch-project.js]: We didn't do this before though, right? Does Apollo maybe have a more generic way to do this e.g. in a link? So we have a SentryErrorReportingLink which enables us having all errors also from the app to be reported? Can be a follow up I just scratched my head cause we added this here now.

File: packages/application-shell/src/components/inject-reducers/inject-reducers.js (hunk 14 above)
Comment (tdeekens) [reply packages/application-shell/src/components/inject-reducers/inject-reducers.js]: Prolly a find and replace gone rogue.

File: packages/application-shell/src/components/navbar/navbar.js (hunk 2 above)
Comment (tdeekens) [reply packages/application-shell/src/components/navbar/navbar.js]: I agree and appriciate this. However, I would also argue that we should split updating libs and  refactoring code. So we only change as little as possible in dep updates and keep such changes for other PRs. I am just arguing for this in the future as we had quite a few regressions in the last weeks.

File: packages/application-shell/src/hooks/use-applications-menu/use-applications-menu.js [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+const defaultApiUrl = window.location.origin;
+
+const defaultConfig = {
+  queryOptions: {},
+  skipRemoteQuery: false,
+  options: {},
+};
+
+const useApplicationsMenu = (config = {}) => {
+  const mergedConfig = { ...defaultConfig, ...config };
```
Comment (tdeekens) [inline packages/application-shell/src/hooks/use-applications-menu/use-applications-menu.js]: Nit: `shallowlyMergedConfig` do not give false impresions.

File: packages/application-shell/src/hooks/use-applications-menu/use-applications-menu.js [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+      if (devConfig.menuLoader) {
+        devConfig.menuLoader().then(setMenu);
+      }
+    }
+  }, [getDevConfig, mergedConfig.skipRemoteQuery]);
+
+  // Prepare to fetch config.
+  // We set up the apollo query as lazy, in order to execute it conditionally
+  // since hooks cannot be defined conditionally.
+  const environment = useApplicationContext(context => context.environment);
```
Comment (tdeekens) [inline packages/application-shell/src/hooks/use-applications-menu/use-applications-menu.js]: Do we really need the whole env or just 1-2 pieces of it?

File: packages/application-shell/src/hooks/use-applications-menu/use-applications-menu.js (hunk 7 above)
Comment (tdeekens) [reply packages/application-shell/src/hooks/use-applications-menu/use-applications-menu.js]: Nice! Stumbled over this for a minute. Should we maybe alias `called` to `hasApplicationsMenuQueryBeenCalled`. I don't know, would just've helped me. I think it's something which will freak me out with Apollo and hooks is that we'll have these super gernic `data` and `called` variable names around.

File: packages/jest-preset-mc-app/setup-tests.js [hunk 19]
```diff
@@ -8,10 +8,18 @@ global.window.app = {
   mcApiUrl: 'http://localhost:8080',
 };
 
+const shouldIgnoreWarnings = (...messages) =>
```
Comment (tdeekens) [inline packages/jest-preset-mc-app/setup-tests.js]: Maybe `silence` is a better word. Not sure.

File: packages/application-shell/src/components/fetch-project/fetch-project.js (hunk 12 above)
Comment (emmenko) [reply packages/application-shell/src/components/fetch-project/fetch-project.js]: Yeah that’s one of those “uncertain topics” that we have. In general we don’t always handle those kind of errors, at least not always and not consistently.
We could have an error link I suppose yes, although I’m not sure we want to report all errors. I think I would prefer to do this for now, on a query-by-query basis.

PS: I added this here because I found out or realized that while I was working on this PR.

Comment (tdeekens) [reply packages/application-shell/src/components/fetch-project/fetch-project.js]: Sure. I am not generally arguing against it. Just saying that maybe we can encapsulate it more and offer it through a link. Maybe an opt-in or opt-out thing through a reportError prop on each query or mutation.

File: packages/application-shell/src/components/navbar/navbar.js (hunk 2 above)
Comment (emmenko) [reply packages/application-shell/src/components/navbar/navbar.js]: Yeah sure, I agree 👌

Just for the record, those refactorings were part of the experimental branch of using the Apollo hooks 2 months ago or something.

File: packages/application-shell/src/hooks/use-applications-menu/use-applications-menu.js (hunk 18 above)
Comment (emmenko) [reply packages/application-shell/src/hooks/use-applications-menu/use-applications-menu.js]: True, we can select the values one by one

File: packages/application-shell/src/hooks/use-applications-menu/use-applications-menu.js (hunk 7 above)
Comment (emmenko) [reply packages/application-shell/src/hooks/use-applications-menu/use-applications-menu.js]: Sure thing, we can alias them if you prefer so

**Comment by emmenko** [comment]
@tdeekens anything else left from your side?

**Comment by tdeekens** [comment]
Not really. It's too general to apply here.

1. Let's not change too much in version updates in the same PR.
   - The argument "it's been done in an older PR" is fine but also odd.
2. We should think of a better `SentryErrorLink` in the future

**Comment by tdeekens** [comment]
Did you require another review by @amine-benselim or not @emmenko?

**Comment by emmenko** [comment]
Hmm no I didn't

**Comment by amine-benselim** [comment]
no he didnt. I was reviewing it while he merged it. so I just continued doing so to know what changed anyway.

File: packages/application-shell/src/components/navbar/navbar.spec.js [hunk 20]
```diff
@@ -49,27 +49,27 @@ const createTestProps = props => ({
...
-  projectExtensionsQuery: {
-    projectExtension: {
-      id: 'p1',
-      applications: [
-        { id: 'p1-a1', navbarMenu: createTestMenuConfig('channels') },
-      ],
-    },
-  },
-  onMenuItemClick: jest.fn(),
+  // areProjectExtensionsEnabled: true,
```
[the remark this answers is not in the document]
Comment (emmenko) [reply packages/application-shell/src/components/navbar/navbar.spec.js]: We don't use HOCs anymore, so they are not necessary.

TL;DR: we need to rewrite the tests to RTL 😅
